### DISCUSSION TITLE
Where do the funds come from in InteractionsTest.t.sol in Lesson 7?

### DISCUSSION THREAD

**Discussion by ivanmetrikin** [Cyfrin/foundry-full-course-cu#504] [Q&A] [answered]
In Lesson 7, the interactions test contract can be reduced to the following

```solidity
contract InteractionsTest is Test {
    FundMe fundMe;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function testUserCanFundAndOwnerWithdraw() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundMe));

        assert(address(fundMe).balance == 0);
    }
}
```
The script being tested is as follows:

```solidity
contract FundFundMe is Script {
    uint256 SEND_VALUE = 0.1 ether;

    function fundFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
    }
}

contract WithdrawFundMe is Script {
    function withdrawFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).withdraw();
        vm.stopBroadcast();
    }
}
```

What puzzles me is that this test `testUserCanFundAndOwnerWithdraw` passes without error, even though we don't prank a user anywhere with `vm.prank()` and don't distribute any ETH with `vm.deal`. What is happening here? Where do the funds for `SEND_VALUE` come from?

I would expect this test to fail/revert, as per the previously written unit test https://github.com/Cyfrin/foundry-fund-me-f23/blob/main/test/unit/FundMeTest.t.sol#L38. However, the 0.1 ether is seemingly appearing out of thin air in the `fundFundMe` function of the `FundFundMe` contract

**Comment by shikhar229169** [comment, marked as answer]
In your test ```testUserCanFundAndOwnerWithdraw```, we are actually calling the fundFundMe function of the FundFundMe interactions contract, and in there we are using broadcast. 
So, where ever we add broadcast the following transaction is performed on behalf of the address, for which we passed the private key in the command using --private-key arg, but while running test we are actually not required to pass the private key, it is automatically selected by foundry and the account also have thousands of fake ETH.

```solidity
function fundFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
}
```

In the above function where we are calling fund on FundMe contract it is not called by FundFundMe contract, as we have set it within broadcast so it is called on behalf of that address I described above and the test passed.

Hope it is clear to you.
And you can also do some console.log to test all these stuffs which will make it more clear.

**Reply by ivanmetrikin** [reply]
Thanks @shikhar229169, you've nailed it with this line: 'but while running test we are actually not required to pass the private key, it is automatically selected by foundry and the account also have thousands of fake ETH.'

I only wish it had been made a little bit clearer in the lesson
